lib/prog.ml

```diff
@@ -786,6 +788,575 @@ end = struct
+
+  type pkt = buf [@@deriving sexp,yojson]
+  type pkt_out = buf * buf [@@deriving sexp,yojson]
+
+  type entries = Table.pre_entry list
+
+  type vsets = Match.t list list
+
+  type ctrl = entries * vsets
+
+  type loc = string [@@deriving sexp, yojson]
+
+  type value =
+    | VNull
+    | VBool of bool
+    | VInteger of Bigint.t
+    | VBit of 
+        { w : Bigint.t;
+          v : Bigint.t; }
+    | VInt of 
+        { w : Bigint.t;
+          v : Bigint.t; }
+    | VVarbit of 
+        { max : Bigint.t;
+          w : Bigint.t;
+          v : Bigint.t; }
+    | VString of string
+    | VTuple of value list
+    | VRecord of (string * value) list
+    | VSet of set
+    | VError of string
+    | VMatchKind of string
+    | VFun of 
+        { params : TypeParameter.t list;
+          body : Block.t; }
+    | VBuiltinFun of 
+        { name : string;
+          caller : lvalue; }
+    | VAction of 
+        { params : TypeParameter.t list;
+          body : Block.t; }
+    | VStruct of 
+        { fields : (string * value) list; }
+    | VHeader of 
+        { fields : (string * value) list;
+          is_valid : bool }
+    | VUnion of 
+        { valid_header : value;
+          valid_fields : (string * bool) list; }
+    | VStack of 
+        { headers : value list;
+          size : Bigint.t;
+          next : Bigint.t; }
+    | VEnumField of 
+        { typ_name : string;
+          enum_name : string; }
+    | VSenumField of 
+        { typ_name : string;
+          enum_name : string;
+          v : value; }
+    | VRuntime of 
+        { loc : loc;
+          obj_name : string; }
+    | VParser of vparser
+    | VControl of vcontrol
+    | VPackage of 
+        { decl : Declaration.t;
+          args : (string * value) list; }
+    | VTable of vtable
+    | VExternFun of 
+        { name : string;
+          caller : (loc * string) option; }
+    [@@deriving sexp,yojson]
+  
+  and vparser = {
+    pvs : (string * value) list;
+    pparams : TypeParameter.t list;
+    plocals : Declaration.t list;
+    states : Parser.state list;
+  }
+  [@@deriving sexp,yojson]
+  
+  and vcontrol = {
+    cvs : (string * value) list;
+    cparams : TypeParameter.t list;
+    clocals : Declaration.t list;
+    apply : Block.t;
+  }
+  [@@deriving sexp,yojson]
+  
+  and vtable = {
+    name : string;
+    keys : value list;
+    actions : Table.action_ref list;
+    default_action : Table.action_ref;
+    const_entries : (set * Table.action_ref) list;
+  }
+  [@@deriving sexp,yojson]
+  
+  and lvalue =
+    | LName of
+      { name : Types.name;
+        typ : Type.t; }
+    | LMember of 
+      { expr : lvalue;
+        name : string;
+        typ : Type.t }
+    | LBitAccess of 
+        { expr : lvalue;
+          msb : Bigint.t;
+          lsb : Bigint.t;
+          typ : Type.t }
+    | LArrayAccess of 
+        { expr : lvalue;
+          idx : value;
+          typ : Type.t; }
+  [@@deriving sexp,yojson]
```

Comment:
Instead of giving a typ field to every lvalue, make this datatype the pre_lvalue and make lvalue = {val: lvalue; typ: Typed.Type.t} like we do for Expression. Then we can get rid of the lvalue_typ function in here.

Reply:
done